drivertest/commontest/common.go

```diff
@@ -0,0 +1,284 @@
+	defer copyist.Open(t).Close()
+
+	// Open database.
+	db, err := sql.Open("copyist_"+driverName, dataSourceName)
+	require.NoError(t, err)
+
+	rows, err := db.Query("SELECT name FROM customers WHERE id=$1", 1)
+	require.NoError(t, err)
+	defer rows.Close()
+
+	for rows.Next() {
+		var name string
+		rows.Scan(&name)
+		require.Equal(t, "Andy", name)
+	}
+}
+
+// TestInsert inserts a row and ensures that it's been committed.
+func TestInsert(t *testing.T, driverName, dataSourceName string) {
+	defer copyist.Open(t).Close()
+
+	// Open database.
+	db, err := sql.Open("copyist_"+driverName, dataSourceName)
+	require.NoError(t, err)
+
+	res, err := db.Exec("INSERT INTO customers VALUES ($1, $2)", 4, "Joel")
+	require.NoError(t, err)
+
+	affected, err := res.RowsAffected()
+	require.NoError(t, err)
+	require.Equal(t, int64(1), affected)
+
+	rows, err := db.Query("SELECT COUNT(*) FROM customers")
+	require.NoError(t, err)
+	defer rows.Close()
+
+	for rows.Next() {
+		var cnt int
+		rows.Scan(&cnt)
+		require.Equal(t, 4, cnt)
+	}
+}
+
+// TestDataTypes queries data types that are interesting for the SQL driver.
+func TestDataTypes(t *testing.T, driverName, dataSourceName string) {
+	defer copyist.Open(t).Close()
+
+	// Open database.
+	db, err := sql.Open("copyist_"+driverName, dataSourceName)
+	require.NoError(t, err)
+
+	// Construct table with many data types.
+	res, err := db.Exec(`
+		CREATE TABLE datatypes
+		(i INT, s TEXT, tz TIMESTAMPTZ, t TIMESTAMP, b BOOL,
+		 by BYTES, f FLOAT, d DECIMAL, fa FLOAT[], u UUID)
+	`)
+	require.NoError(t, err)
+
+	_, err = db.Exec(`
+		INSERT INTO datatypes VALUES
+			(1, 'foo' || CHR(9) || CHR(10) || ' ,]', '2000-01-01T10:00:00Z', '2000-01-01T10:00:00Z',
+			 true, 'ABCD', 1.1, 100.1234, ARRAY(1.1, 1.2345678901234567),
+			 '8B78978B-7D8B-489E-8CA9-AC4BDC495A82'),
+			(2, '', '2000-02-02T11:11:11-08:00', '2000-02-02T11:11:11-08:00', false,
+			 '', -1e10, -0.0, ARRAY(), '00000000-0000-0000-0000-000000000000')
+	`)
+	require.NoError(t, err)
+
+	affected, err := res.RowsAffected()
+	require.NoError(t, err)
+	require.Equal(t, int64(0), affected)
+
+	var out DataTypes
+	rows, err := db.Query("SELECT i, s, tz, t, b, by, f, d, fa, u FROM datatypes")
+	require.NoError(t, err)
+
+	rows.Next()
+	require.NoError(t, rows.Scan(
+		&out.Int, &out.Str, &out.TimeZ, &out.Time, &out.Bool, &out.Bytes,
+		&out.Flt, &out.Dec, &out.FltArr, &out.Uuid))
+	require.Equal(t, DataTypes{
+		Int: 1, Str: "foo\t\n ,]", TimeZ: parseTime("2000-01-01T10:00:00Z", driverName),
+		Time: parseTime("2000-01-01T10:00:00+00:00", driverName), Bool: true,
+		Bytes: []byte{'A', 'B', 'C', 'D'}, Flt: 1.1, Dec: "100.1234",
+		FltArr: "{1.1,1.2345678901234567}", Uuid: []byte("8b78978b-7d8b-489e-8ca9-ac4bdc495a82"),
+	}, out)
+
+	rows.Next()
+	require.NoError(t, rows.Scan(
+		&out.Int, &out.Str, &out.TimeZ, &out.Time, &out.Bool, &out.Bytes,
+		&out.Flt, &out.Dec, &out.FltArr, &out.Uuid))
+	require.Equal(t, DataTypes{
```

Comment:
Huh, I'm surprised this works. Normally you can't do comparison on structs if the struct contains a slice such as `[]byte`. Must be some magic going on in the `require` package.